pkg/dns/dns_test.go

```diff
@@ -0,0 +1,377 @@
+	return &service
+}
+
+func newEndpoints(service *kapi.Service, subsets ...kapi.EndpointSubset) *kapi.Endpoints {
+	endpoints := kapi.Endpoints{
+		ObjectMeta: service.ObjectMeta,
+		Subsets:    []kapi.EndpointSubset{},
+	}
+
+	for _, subset := range subsets {
+		endpoints.Subsets = append(endpoints.Subsets, subset)
+	}
+	return &endpoints
+}
+
+func newSubsetWithOnePort(portName string, port int32, ips ...string) kapi.EndpointSubset {
+	subset := newSubset()
+	subset.Ports = append(subset.Ports, kapi.EndpointPort{Port: port, Name: portName, Protocol: "TCP"})
+	for _, ip := range ips {
+		subset.Addresses = append(subset.Addresses, kapi.EndpointAddress{IP: ip})
+	}
+	return subset
+}
+
+func newSubsetWithTwoPorts(portName1 string, portNumber1 int32, portName2 string, portNumber2 int32, ips ...string) kapi.EndpointSubset {
+	subset := newSubsetWithOnePort(portName1, portNumber1, ips...)
+	subset.Ports = append(subset.Ports, kapi.EndpointPort{Port: portNumber2, Name: portName2, Protocol: "TCP"})
+	return subset
+}
+
+func newSubset() kapi.EndpointSubset {
+	subset := kapi.EndpointSubset{
+		Addresses: []kapi.EndpointAddress{},
+		Ports:     []kapi.EndpointPort{},
+	}
+	return subset
+}
+
+func assertSRVForHeadlessService(t *testing.T, kd *KubeDNS, s *kapi.Service, e *kapi.Endpoints) {
+	for _, subset := range e.Subsets {
+		for _, port := range subset.Ports {
+			records, err := kd.Records(getSRVFQDN(kd, s, port.Name), false)
+			require.NoError(t, err)
+			assertRecordPortsMatchPort(t, port.Port, records)
+			assertCNameRecordsMatchEndpointIPs(t, kd, subset.Addresses, records)
+		}
+	}
+}
+
+func assertDNSForHeadlessService(t *testing.T, kd *KubeDNS, e *kapi.Endpoints) {
+	records, err := kd.Records(getEndpointsFQDN(kd, e), false)
+	require.NoError(t, err)
+	endpoints := map[string]bool{}
+	for _, subset := range e.Subsets {
+		for _, endpointAddress := range subset.Addresses {
+			endpoints[endpointAddress.IP] = true
+		}
+	}
+	assert.Equal(t, len(endpoints), len(records))
+	for _, record := range records {
+		_, found := endpoints[record.Host]
+		assert.True(t, found)
+	}
+}
+
+func assertRecordPortsMatchPort(t *testing.T, port int32, records []skymsg.Service) {
+	for _, record := range records {
+		assert.Equal(t, port, int32(record.Port))
+	}
+}
+
+func assertCNameRecordsMatchEndpointIPs(t *testing.T, kd *KubeDNS, e []kapi.EndpointAddress, records []skymsg.Service) {
+	endpoints := map[string]bool{}
+	for _, endpointAddress := range e {
+		endpoints[endpointAddress.IP] = true
+	}
+	assert.Equal(t, len(e), len(records), "unexpected record count")
+	for _, record := range records {
+		_, found := endpoints[getIPForCName(t, kd, record.Host)]
+		assert.True(t, found, "Did not endpoint with address:%s", record.Host)
+	}
+}
+
+func getIPForCName(t *testing.T, kd *KubeDNS, cname string) string {
+	records, err := kd.Records(cname, false)
+	require.NoError(t, err)
+	assert.Equal(t, 1, len(records), "Could not get IP for CNAME record for %s", cname)
```

Comment:
"Could not get the expected number of records for %s" may be? len(records) might be > 1 and we still get this assertion failure.